utils/simplecache/cache.go

```diff
@@ -0,0 +1,210 @@
+package simplecache
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	fs "github.com/ARM-software/golang-utils/utils/filesystem"
+	"github.com/ARM-software/golang-utils/utils/parallelisation"
+)
+
+const (
+	_DefaultGCPeriod = 10 * time.Minute
+	_DefaultTTL      = 2 * time.Hour
+)
+
+type CacheEntry struct {
+	path         string
+	originalFS   fs.FS
+	originalpath string
+	expiration   time.Time
+}
+
+type Cache struct {
+	entries map[string]*CacheEntry
+	cfg     *Config
+	stopGC  context.CancelFunc
+	mu      sync.Mutex
+	closed  bool
+}
+
+func (c *Cache) gc(ctx context.Context, _ time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := time.Now()
+	for id, entry := range c.entries {
+		if now.After(entry.expiration) {
+			if err := c.cfg.cachefs.RemoveWithContext(ctx, entry.path); err == nil {
+				delete(c.entries, id)
+			}
+		}
+	}
+}
+
+func (c *Cache) isClosed() error {
+	if c.closed {
+		return commonerrors.New(commonerrors.ErrForbidden, "cache is closed")
+	}
+
+	return nil
+}
+
+func (c *Cache) Contains(ctx context.Context, id string) (bool, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if err := c.isClosed(); err != nil {
+		return false, err
+	}
+
+	_, exists := c.entries[id]
+	return exists, nil
+}
+
+func (c *Cache) Restore(ctx context.Context, id string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if err := c.isClosed(); err != nil {
+		return err
+	}
+
+	entry, exists := c.entries[id]
+	if !exists {
+		return commonerrors.Newf(commonerrors.ErrNotFound, "cache entry for %s not found", id)
+	}
+
+	srcPath := fs.FilePathJoin(c.cfg.cachefs, c.cfg.cachePath, fs.FilePathBase(c.cfg.cachefs, entry.path))
+	err := fs.CopyBetweenFS(ctx, c.cfg.cachefs, srcPath, entry.originalFS, entry.originalpath)
```

Comment:
this is not atomic and so there could be system trying to use the file while being copied

Reply:
one way to solve this is to store the file in a temp file in the same filesystem and then mv it to the new name when done. the cache should only report it has an entry if the entry is usable not in transient state